<a href="https://colab.research.google.com/github/GAPatterson97/sales_predictions/blob/main/sales_predictions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sales Predictions
*   Glen A Patterson
*   07/27/2022

In [178]:
#import Libraries
import pandas as pd

In [179]:
# read url csv file and store as df
df = pd.read_csv('https://docs.google.com/spreadsheets/d/e/2PACX-1vSeW9ale6iXr0r93iHDDBj87LxoqaxrzOMtmn-c2HDHWNp-PevBYvkGfIRIMk93-2djm2r7DyBoMuv2/pub?gid=481181311&single=true&output=csv')

In [180]:
df.head(12) #read the first 12 rows in df (sales predictions)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0            FDA15        9.300          Low Fat         0.016047   
1            DRC01        5.920          Regular         0.019278   
2            FDN15       17.500          Low Fat         0.016760   
3            FDX07       19.200          Regular         0.000000   
4            NCD19        8.930          Low Fat         0.000000   
5            FDP36       10.395          Regular         0.000000   
6            FDO10       13.650          Regular         0.012741   
7            FDP10          NaN          Low Fat         0.127470   
8            FDH17       16.200          Regular         0.016687   
9            FDU28       19.200          Regular         0.094450   
10           FDY07       11.800          Low Fat         0.000000   
11           FDA03       18.500          Regular         0.045464   

                Item_Type  Item_MRP Outlet_Identifier  \
0                   Dairy  249.8092            OUT049   
1             Soft Drinks   48.2692            OUT018   
2                    Meat  141.6180            OUT049   
3   Fruits and Vegetables  182.0950            OUT010   
4               Household   53.8614            OUT013   
5            Baking Goods   51.4008            OUT018   
6             Snack Foods   57.6588            OUT013   
7             Snack Foods  107.7622            OUT027   
8            Frozen Foods   96.9726            OUT045   
9            Frozen Foods  187.8214            OUT017   
10  Fruits and Vegetables   45.5402            OUT049   
11                  Dairy  144.1102            OUT046   

    Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                        1999      Medium               Tier 1   
1                        2009      Medium               Tier 3   
2                        1999      Medium               Tier 1   
3                        1998         NaN               Tier 3   
4                        1987        High               Tier 3   
5                        2009      Medium               Tier 3   
6                        1987        High               Tier 3   
7                        1985      Medium               Tier 3   
8                        2002         NaN               Tier 2   
9                        2007         NaN               Tier 2   
10                       1999      Medium               Tier 1   
11                       1997       Small               Tier 1   

          Outlet_Type  Item_Outlet_Sales  
0   Supermarket Type1          3735.1380  
1   Supermarket Type2           443.4228  
2   Supermarket Type1          2097.2700  
3       Grocery Store           732.3800  
4   Supermarket Type1           994.7052  
5   Supermarket Type2           556.6088  
6   Supermarket Type1           343.5528  
7   Supermarket Type3          4022.7636  
8   Supermarket Type1          1076.5986  
9   Supermarket Type1          4710.5350  
10  Supermarket Type1          1516.0266  
11  Supermarket Type1          2187.1530

## 1.   How many rows and columns?

In [181]:
df.shape #(rows,columns)

(8523, 12)

**- There are 8,523 rows and 12 columns**

## 2.   What are the datatypes of each variable?

In [182]:
df.info() #List information about Dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


**- There is four float64, one int64, and seven object datatypes listed above**

## 3.   Are there duplicates? If so, drop any duplicates.

In [183]:
df.duplicated().sum()  #find the sum of all duplicated entries

0

**- There is no duplicated entries in this dataframe**

## 4.   Identify missing values.

In [184]:
df.info() #List information about Dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


**- We have missing data from columns "Item_Weight" and "Outlet_Size"**
*   Item_Weight has 7,060 out of 8,523 entries = 1,463 missing entries. approx. 17.2%
*   Outlet_Size has 6,113 out of 8,523 entries = 2,410 missing entries. approx. 28.3%



## 5.   Decide on how to address the missing values and do it! 

(This requires your judgement, so explain your choice).

**There is to much data that would be lossed if we just remove the items with rows of missing data**
*   For the Item_Weight we will fill in missing data with the average all the items listed weights.
*   For the Outlet_Size we will fill in missing data with the most common type of all the stores. 

      (In terms of ground area covered: "Small", "Medium" , "High")

In [185]:
Item_Weight_Avg = df['Item_Weight'].mean().round(3)
Item_Weight_Avg

12.858

**- The average listed weight for all the items is 12.858**

In [186]:
df['Item_Weight'].fillna(Item_Weight_Avg, inplace=True) #Fill in all NaN with average item weight

In [187]:
df['Outlet_Size'].nunique() #the number of unique store types

3

In [188]:
Small_Store = (df['Outlet_Size'] == 'Small').sum()  #Sum of Small stores
Medium_Store = (df['Outlet_Size'] == 'Medium').sum() #Sum of Medium stores
High_Store = (df['Outlet_Size'] == 'High').sum() #Sum of High stores
print(f'Small store type: {Small_Store}   Medium store type: {Medium_Store}   High store type: {High_Store}')

Small store type: 2388   Medium store type: 2793   High store type: 932


**- The most common store is Medium size store so we will fill in all missing data with that store type.**

In [189]:
df['Outlet_Size'].fillna('Medium', inplace=True)

## 6.   Confirm that there are no missing values after addressing them.

In [190]:
df.info() #List information about Dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                8523 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                8523 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


**- From list above we can see that there is no missing data**
*   All columns have 8,523 out of 8,523 entries

## 7.   Find and fix any inconsistent categories of data (example: fix cat, Cat, and cats so that they are consistent)

In [191]:
df['Item_Fat_Content'].unique() #find the inconsistent data entered for column Item_Fat_Content

array(['Low Fat', 'Regular', 'low fat', 'LF', 'reg'], dtype=object)

**- Fixing all to just "Low Fat" and "Regular"**

In [192]:
for idx, row in df.iterrows(): # looping Item_Fat_Content correcting "LF", "low fat", and "reg"
  if (df.loc[idx, 'Item_Fat_Content'] == 'LF') | (df.loc[idx, 'Item_Fat_Content'] == 'low fat'):
    df.loc[idx, 'Item_Fat_Content'] = 'Low Fat'
  if df.loc[idx, 'Item_Fat_Content'] == 'reg':
    df.loc[idx, 'Item_Fat_Content'] = 'Regular'

**- Checking for the correction in data**

In [193]:
df['Item_Fat_Content'].unique() #check correction made on column Item_Fat_Content

array(['Low Fat', 'Regular'], dtype=object)

In [194]:
df['Item_Type'].unique() #find the inconsistent data entered for column Item_Type

array(['Dairy', 'Soft Drinks', 'Meat', 'Fruits and Vegetables',
       'Household', 'Baking Goods', 'Snack Foods', 'Frozen Foods',
       'Breakfast', 'Health and Hygiene', 'Hard Drinks', 'Canned',
       'Breads', 'Starchy Foods', 'Others', 'Seafood'], dtype=object)

In [195]:
df['Outlet_Identifier'].unique() #find the inconsistent data entered for column Outlet_Identifier

array(['OUT049', 'OUT018', 'OUT010', 'OUT013', 'OUT027', 'OUT045',
       'OUT017', 'OUT046', 'OUT035', 'OUT019'], dtype=object)

In [196]:
df['Outlet_Establishment_Year'].unique() #find the inconsistent data entered for column Outlet_Establishment_Year

array([1999, 2009, 1998, 1987, 1985, 2002, 2007, 1997, 2004])

In [197]:
df['Outlet_Location_Type'].unique() #find the inconsistent data entered for column Outlet_Location_Type

array(['Tier 1', 'Tier 3', 'Tier 2'], dtype=object)

In [198]:
df['Outlet_Type'].unique() #find the inconsistent data entered for column Outlet_Type

array(['Supermarket Type1', 'Supermarket Type2', 'Grocery Store',
       'Supermarket Type3'], dtype=object)

**- No other inconsistent data found!  Only Item_Fat_Content was inconsistent**

## 8.   For any numerical columns, obtain the summary statistics of each (min, max, mean)

In [199]:
df.describe() #statistics for all the numerical coulmns

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  8523.000000      8523.000000  8523.000000                8523.000000   
mean     12.857706         0.066132   140.992782                1997.831867   
std       4.226124         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       9.310000         0.026989    93.826500                1987.000000   
50%      12.858000         0.053931   143.012800                1999.000000   
75%      16.000000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

**- Above is the summary of the statistics for each of the columns**